### PR TITLE
TestPMD workload 

### PR DESCRIPTION
**Pull request by mukrishn**
This PR is part -1 of testpmd workload, this creates a standalone testpmd pod with SR-IOV VF up and listening for traffic. User can manually run a traffic generator(on a VM/Baremetal) and test the throughput.
Part-2 will be, including a traffic generation(trex) pod along with this - Included in this PR

### CODE REVIEW

File: docs/testpmd.md [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+  spoofChk: "on"
+  trust: "on"
+  resourceName: intelnics
+  networkNamespace: my-ripsaw
+
+```
+
+### Sample PAO configuration
+
+The CPU and socket information might vary, this is defined based on the recommendation [here](https://docs.google.com/document/d/1wlT8z-QNzibgmNZnw5M3QNbwBcE0bJEcjWwh_BoA1Ig/edit?usp=sharing)
```
Comment (jtaleric) [inline docs/testpmd.md:60]: Doubtfully this doc is public, just remove it. 

`The CPU and socket information might vary, use the below as an example`

File: docs/testpmd.md [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    args:
+      privileged: true
+      pin: false
+      pin_testpmd: "node-0"      
+      networks:
+        testpmd:
+          - name: testpmd-sriov-network
+            count: 2  # Interface count
+            mac:
+              - 60:60:00:f4:47:06
```
Comment (jtaleric) [inline docs/testpmd.md:111]: Where does the user determine the MAC?

**Review by jtaleric — COMMENTED** [review]
Since this doesn't launch a workload, how does part2 come into play? Is part2 adding to the testpmd workload? ie testpmd+trafficgen for part2? Curious how you envision this evolving?

File: docs/testpmd.md (hunk 2 above)
Comment (mukrishn) [reply docs/testpmd.md:111]: These are optional, random mac input from user for the SR-IOV VF. If not given script generates it [here](https://github.com/cloud-bulldozer/benchmark-operator/pull/510/files#diff-8faf8cd1f3463c3da5f4e18c423fcb7e1922806af35cd54c5e8519b72a98c33aR7). User must keep track of these mac for traffic generation. 
Where as `peer_mac` is traffic generator interface mac.
I will update this in readme.

**Comment by mukrishn** [comment]
Yes, planning to run another job with trex as part of the same workload, it would look like testpmd(server) + trafficgen(client) pair
Trying to bring in a simplest version of this https://github.com/rh-nfv-int/trex-operator or trafficgen if it supports pod level implementation.

**Review by jtaleric — COMMENTED** [review]
The other issue I see here is testing the solution. We will have to depend on you and @smalleni to execute and verify that this is working.

**Comment by mukrishn** [comment]
@jtaleric agree, this PR cannot be easily verified without a proper traffic generator set up, let this be open for few days and close this along with part-2 or I can open a fresh PR altogether.

**Review by smalleni — COMMENTED** [review]
@mukrishn You might need to rebase, since ` charts/benchmark-operator/crds/ripsaw_v1alpha1_ripsaw_crd.yaml ` should be a symlink. I see your commit as adding that again.

File: docs/testpmd.md [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+      "gateway": "10.57.1.1"
+    }
+  spoofChk: "on"
+  trust: "on"
+  resourceName: intelnics
+  networkNamespace: my-ripsaw
+
+```
+
+### Sample PAO configuration
```
Comment (smalleni) [inline docs/testpmd.md:58]: Please add note about waiting for MCP to be updated after applying Sr-IOV

**Review by smalleni — APPROVED** [review]
Works flawlessly, just need to get the trex image pushed to cloud-bulldozer if not already done.

Output: https://gist.github.com/smalleni/5bb606b2285cb2706f1de0f75a7c000d

I do think in a later commit we need to introduce the trafficgen/binary search methodology for max pps with given loss %

**Comment by mukrishn** [comment]
@jtaleric this PR is verified by @smalleni already, I think this is good to merge unless you have some comments or thoughts.

**Review by jtaleric — APPROVED** [review]
LGTM!

**Comment by sarahbx** [comment]
👍 LGTM

File: docs/testpmd.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
+# TestPMD 
+
+[TestPMD](https://doc.dpdk.org/guides/testpmd_app_ug/index.html) is anapplication  used to test the DPDK in a packet forwarding mode, here this application run inside a pod with SRIOV VF.
```
Comment (HughNhan) [inline docs/testpmd.md:3]: "is anapplication" -> "is an application"

File: docs/testpmd.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+   nodeSelector:
+      node-role.kubernetes.io/worker: ""
+   realTimeKernel:
+     enabled: false
+   additionalKernelArgs:
+   - nosmt
+   - tsc=reliable
+```
+
+## Running standalone TestPMD
```
Comment (HughNhan) [inline docs/testpmd.md:95]: Should it be "Running containerized TestPMD" ?

File: roles/testpmd/templates/testpmd.yml.j2 [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        args:
+        - "-c"
+        - echo "set promisc all off" >> /tmp/testpmd-cmdline.txt && testpmd -l $(cat /sys/fs/cgroup/cpuset/cpuset.cpus) $(PCI=""; IFS=',' read -r -a NETWORK_ARRAY <<< "$NETWORK_NAME_LIST"; for item in "${NETWORK_ARRAY[@]}"; do IFS='/' read -r -a RES_ARRAY <<< "$item"; NAME="PCIDEVICE_OPENSHIFT_IO_${RES_ARRAY[1]^^}";IFS=',' read -r -a PCI_ARRAY <<< "${!NAME}";for pci_item in "${PCI_ARRAY[@]}"; do PCI+=" -w ${pci_item} ";done;done;echo $PCI) --socket-mem {{ workload_args.socket_memory | default('1024,0') }} -n {{ workload_args.memory_channels | default(4) }} --proc-type auto --file-prefix pg -- --disable-rss --nb-cores={{ workload_args.forwarding_cores | default(4) }} --rxq={{ workload_args.rx_queues | default(1) }} --txq={{ workload_args.tx_queues | default(1) }} --rxd={{ workload_args.rx_descriptors | default(1024) }} --txd={{ workload_args.tx_descriptors | default(1024) }} --auto-start {% for mac in maclist_trex %} --eth-peer={{ loop.index -1 }},{{ mac }} {% endfor %} --forward-mode={{ workload_args.forward_mode | default('mac') }} --stats-period {{ workload_args.stats_period | default(1) }} --cmdline-file /tmp/testpmd-cmdline.txt
```
Comment (HughNhan) [inline roles/testpmd/templates/testpmd.yml.j2:41]: So we always disable rss, or that how tespmd does?

**Review by HughNhan — COMMENTED** [review]
A few minor typos, and a general question about RSS

File: docs/testpmd.md (hunk 5 above)
Comment (mukrishn) [reply docs/testpmd.md:95]: Correct, will update

File: roles/testpmd/templates/testpmd.yml.j2 (hunk 6 above)
Comment (mukrishn) [reply roles/testpmd/templates/testpmd.yml.j2:41]: No, we are optionally disabling rss for a basic load profile. We can make it user configurable with a default value.

**Review by HughNhan — APPROVED** [review]
LGTM

**Review by jtaleric — APPROVED** [review]

**Pull request merged**